### ISSUE TITLE
Feature Request: Config option for specifying other directories from which data/resource packs can be loaded

### ISSUE THREAD

**Issue by yungnickyoung** [yungnickyoung/Packie#33]

**Comment by Fyoncle** [comment]
@yungnickyoung That is a must have because developers packs do not show as embedded and they get 0 download, and that's the worst thing can happen. Will this be implemented any soon?

**Comment by yungnickyoung** [comment]
Any modpacks using Paxi to load packs that exist on CurseForge should be manually marking them as embedded dependencies. Is that not happening?

**Comment by Fyoncle** [comment]
> Any modpacks using Paxi to load packs that exist on CurseForge should be manually marking them as embedded dependencies. Is that not happening?

I don't know about CurseForge, but Modrinth does not. Because Modrinth detects them from corresponding folders of theirs, and if they don't exist there all the resource pack devs get 0 download.

**Comment by Fyoncle** [comment]
You can manually go on putting them as embedded which if you have much resource packs like me around 30, it will be a pain for every release. But i believe that it will download them twice if you do that then, not sure.

**Comment by yungnickyoung** [comment]
If Modrinth doesn't have support for manually adding packs then I'll have to look into adding this soon. But CF definitely does given how many modpacks Paxi is used in, and I don't think there are any issues with duplicate downloads. Also, you can mark the packs as global included dependencies and they should be automatically added to every release, so there's no need to manually edit each release.

**Comment by Fyoncle** [comment]
> global included dependencies

Wdym by global included dependencies? And by duplicate downloads I meant that if its embedded i think it also downloads it once again to the resource pack folder alongside the paxi folder, so how does that even work? Because if it didnt download it again if it was embedded but still they get downloads, that would be a massive error in monetization system, allowing people to add embedded stuff of theirs even tho if its not in the modpack and make more downloads.

**Comment by yungnickyoung** [comment]
> Wdym by global included dependencies? 

You can add "Default Relations" to a project, which will then automatically apply to every release.

> And by duplicate downloads I meant that if its embedded i think it also downloads it once again to the resource pack folder alongside the paxi folder, so how does that even work? Because if it didnt download it again if it was embedded but still they get downloads, that would be a massive error in monetization system, allowing people to add embedded stuff of theirs even tho if its not in the modpack and make more downloads.

I thought that if you simply add a project as a dependency manually, either when uploading or as a default relation, then it marks that project as a dependency without downloading anything. You can then separately add the pack to paxi's config folder via an override.

**Comment by Fyoncle** [comment]
> I thought that if you simply add a project as a dependency manually, either when uploading or as a default relation, then it marks

Wdym by override? Like the feature that we waiting for?

**Comment by yungnickyoung** [comment]
No, just like how modpacks can have an "overrides" folder for including things like default configs, resourcepacks, third party mods, etc.

**Comment by Fyoncle** [comment]
> No, just like how modpacks can have an "overrides" folder for including things like default configs, resourcepacks, third party mods, etc.

But still that feature would be good so that it would be automated instead of manually adding those.

**Comment by yungnickyoung** [comment]
Gotcha. I remember I tried a while ago and something about it was particularly challenging since Paxi modifies pack load order, but I'll try to give it another shot soon.

**Comment by Fyoncle** [comment]
> Gotcha. I remember I tried a while ago and something about it was particularly challenging since Paxi modifies pack load order, but I'll try to give it another shot soon.

Thanks! It would be great if you drop a comment here when It's done because I don't wanna make another release of my modpack without devs getting what they deserve 😔

**Comment by yungnickyoung** [comment]
Btw, I assume this conversation mostly applies to resource packs specifically, right? i.e. wanting to automatically load any resource packs placed in the vanilla `resourcepacks` folder

**Comment by Fyoncle** [comment]
> Btw, I assume this conversation mostly applies to resource packs specifically, right? i.e. wanting to automatically load any resource packs placed in the vanilla `resourcepacks` folder

Not automatically, still a load order file but instead of `config > paxi > resourcepacks` i want them to work in `yourmodpackfolder > resourcepacks` so that they also show up as embedded automatically in every release of the modpack.

**Comment by yungnickyoung** [comment]
How about if the resource pack load order JSON is also capable of referencing files in the vanilla resourcepacks folder? For example, if I have a load order json entry called `MyPack.zip`, it would first check the config/paxi/resourcepacks folder for MyPack.zip, and if no matching zip was found, it would then check vanilla resourcepacks.

That way you wouldn't have to manage two different JSON files.

**Comment by Fyoncle** [comment]
> How about if the resource pack load order JSON is also capable of referencing files in the vanilla resourcepacks folder? For example, if I have a load order json entry called `MyPack.zip`, it would first check the config/paxi/resourcepacks folder for MyPack.zip, and if no matching zip was found, it would then check vanilla resourcepacks.
> 
> That way you wouldn't have to manage two different JSON files.

Hmmm that would be a little complicated and unclear, maybe instead of just defining 1 place let there be a config screen that lets you put a path and Paxi looks for that path if it exists, like you can even do `yourmodpackfolder > random-folder > random-folder2 > resourcepack.zip`

**Comment by yungnickyoung** [comment]
Eh I'm not really interested in trying to add a custom directory option atm, it's not really necessary. Just focusing on supporting the vanilla RP directory specifically for the sake of creator attribution like you've mentioned.

And I don't really see how it's complicated, I think it's pretty straightforward to simply state that Paxi looks for RPs in the paxi folder first, and then in the vanilla folder.

**Comment by Fyoncle** [comment]
> Eh I'm not really interested in trying to add a custom directory option atm, it's not really necessary. Just focusing on supporting the vanilla RP directory specifically for the sake of creator attribution like you've mentioned.
> 
> And I don't really see how it's complicated, I think it's pretty straightforward to simply state that Paxi looks for RPs in the paxi folder first, and then in the vanilla folder.

That can be good but what if theres the same file in both directories? Or if we only want one of them instead of both

**Comment by yungnickyoung** [comment]
It would choose the first one found (in the Paxi folder) as an override, and probably log a warning message

**Comment by Fyoncle** [comment]
> It would choose the first one found (in the Paxi folder) as an override, and probably log a warning message

Then that's a great way to fix! Will there be a 1.20.1 port of this too? I hope not only for 1.21.4 xD

**Comment by yungnickyoung** [comment]
Unfortunately it will probably have to stay limited to 1.21.1+. The mod went through a rewrite when porting to 1.21 and I think it would be difficult to backport this addition

**Comment by Fyoncle** [comment]
> Unfortunately it will probably have to stay limited to 1.21.1+. The mod went through a rewrite when porting to 1.21 and I think it would be difficult to backport this addition

Oh, I'm using 1.20.1 so that kind of sounds like a pain. I can understand that it is hard to do so but there is also thousands of people still using 1.20.1 which is the version of my modpack Elysium Days too, so if it was coming to 1.20.1 it would have been great.

Specially when Resource Pack Overrides is the only alternative and its much more complicated than Paxi while your mod is just works the best even if the file is missing, doesn't require a lot of syntaxing, and just well made.

**Comment by Fyoncle** [comment]
If it can't be backported though, can't it be implemented in the current state of 1.20.1 branch?

**Comment by yungnickyoung** [comment]
Sorry, but probably not. I just don't really have the time to maintain my mods on multiple versions right now. And the implementation on 1.20.1 would be much more difficult than 1.21.1 (which is already proving to be tricky)

**Comment by Fyoncle** [comment]
> Sorry, but probably not. I just don't really have the time to maintain my mods on multiple versions right now. And the implementation on 1.20.1 would be much more difficult than 1.21.1 (which is already proving to be tricky)

Oh that is very sad news man, gotcha then.. have a nice day!

**Comment by Fyoncle** [comment]
@yungnickyoung If i can success, can i just replace some mixins of Paxi through my own mod and change resource pack directory to the folder i want? It's called Elysium Days Tweaks, would the license allow me do that?

**Comment by Fyoncle** [comment]
Damn it, even if its not specified in the load order paxi forcefully loads all the packs in the directory, that is a problem

**Comment by yungnickyoung** [comment]
Alright, v5.1.3 for 1.21.1 now supports specifying pathnames in the load_order json files! Pathnames are relative to the base .minecraft folder for that instance, so for example if you want to load the pack MyPack.zip in the vanilla resourcepacks folder, you would add `"resourcepacks/MyPack.zip"` to your resourcepack_load_order.json.

**Comment by Fyoncle** [comment]
> Alright, v5.1.3 for 1.21.1 now supports specifying pathnames in the load_order json files! Pathnames are relative to the base .minecraft folder for that instance, so for example if you want to load the pack MyPack.zip in the vanilla resourcepacks folder, you would add `"resourcepacks/MyPack.zip"` to your resourcepack_load_order.json.

And my friend working on a mod called Paxi Fix that will make it so it uses resource pack folder on 1.20.1

**Comment by Fyoncle** [comment]
![Image](https://github.com/user-attachments/assets/5228f876-cdc8-4115-8ad8-a7b56af9685e)
